app/services/storage.js

```diff
@@ -180,10 +180,37 @@ export class ProfileStorage {
                .then((id) => this.savePlaceIDMapping(url, id));
   }
 
-  // Ordered by last visit, descending.
-  async visited(since = 0, limit = 10) {
+  collectURLs(rows) {
     const out = [];
+    for (const row of rows) {
+      out.push(row.url);
+    }
+    return out;
+  }
 
+  async visitedMatches(substring, since = 0, limit = 10) {
+    // Fetch all places visited, with the latest timestamp for each.
+    // We can't limit here, because we don't know matching URLs.
+    // TODO: inverting this query might work better if `since` is long ago.
+    const sub = `SELECT place, MAX(ts) AS latest
```

Comment:
This'll get reworked a little when we have to string match on more things, and yes, there will be some intersection of in-DB views, in-DB materialized views, and in-memory materialized views.